File: src/backends/pushover.rs
use super::common::{Backend, BackendError, Priority, SendOption};
use async_trait::async_trait;
use serde::{Deserialize, Serialize};

const API_TOKEN: &str = "abughxjjtuofgt89bz21mibut67j5t";
const API_URL: &str = "https://api.pushover.net/1/messages.json";

#[derive(Deserialize, Serialize)]
struct Body {
    token: String,
    user: String,
    title: String,
    message: String,
    device: Option<String>,
    priority: Option<isize>,
    retry: Option<usize>,
    expire: Option<usize>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PushoverConfig {
    pub pushover: PushoverBackend,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PushoverBackend {
    user_key: String,
    device: Option<String>,
    priority: Option<Priority>,
    retry: Option<usize>,
    expire: Option<usize>,
}

#[async_trait]
impl Backend for PushoverBackend {
    async fn send(&self, msg: &str, title: &str, option: &SendOption) -> Result<(), BackendError> {
        let body = &Body {
            token: API_TOKEN.to_string(),
            user: self.user_key.to_string(),
            title: title.to_string(),
            message: msg.to_string(),
            device: match option.pushover_device.clone() {
                Some(device) => Some(device),
                None => match self.device.clone() {
                    Some(device) => Some(device),
                    None => None,
                },
            },
            priority: match &option.pushover_priority {
                Some(priority) => Some(priority_to_pushover_code(priority)),
                None => match &self.priority {
                    Some(priority) => Some(priority_to_pushover_code(priority)),
                    None => None,
                },
            },
            retry: match option.pushover_retry {
                Some(retry) => Some(retry),
                None => match self.retry {
                    Some(retry) => Some(retry),
                    None => None,
                },
            },
            expire: match option.pushover_expire {
                Some(expire) => Some(expire),
                None => match self.expire {
                    Some(expire) => Some(expire),
                    None => None,
                },
            },
        };
        let req = match surf::post(API_URL).body_json(body) {
            Ok(req) => req,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        let res = match req.await {
            Ok(res) => res,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        if res.status() != 200 {
            return Err(BackendError::Message(res.status().to_string()));
        }

        Ok(())
    }
}

pub fn priority_to_pushover_code(priority: &Priority) -> isize {
    match priority {
        Priority::Emergency => 2,
        Priority::High => 1,
        Priority::Normal => 0,
        Priority::Low => -1,
        Priority::Lowest => -2,
    }
}

File: backends/pushover.go
package backends

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

func init() {
	backends["pushover"] = NewPushover()
}

const PUSHOVER_API_TOKEN = "abughxjjtuofgt89bz21mibut67j5t"
const PUSHOVER_API_URL = "https://api.pushover.net/1/messages.json"

type PushoverConfig struct {
	UserKey  string  `mapstructure:"user_key" validate:"required"`
	Device   *string `mapstructure:"device" validate:"omitempty"`
	Priority *string `mapstructure:"priority" validate:"omitempty,oneof=emergency high normal low lowest"`
	Retry    *int    `mapstructure:"retry" validate:"omitempty,min=30"`
	Expire   *int    `mapstructure:"expire" validate:"omitempty,min=0,max=10800"`
}

type Pushover struct {
}

func NewPushover() BackendInterface {
	return &Pushover{}
}

func (*Pushover) GetConfig() interface{} {
	return PushoverConfig{}
}

func (*Pushover) Send(configIface interface{}, title string, message string, status *bool) error {
	config, ok := configIface.(PushoverConfig)
	if !ok {
		return fmt.Errorf("invalid config")
	}

	body := map[string]interface{}{
		"token":   PUSHOVER_API_TOKEN,
		"user":    config.UserKey,
		"title":   title,
		"message": message,
	}
	if config.Device != nil {
		body["device"] = *config.Device
	}
	if config.Priority != nil {
		switch *config.Priority {
		case "emergency":
			body["priority"] = 2
		case "high":
			body["priority"] = 1
		case "normal":
			body["priority"] = 0
		case "low":
			body["priority"] = -1
		case "lowest":
			body["priority"] = -2
		}
	}
	if config.Retry != nil {
		body["retry"] = *config.Retry
	}
	if config.Expire != nil {
		body["expire"] = *config.Expire
	}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return err
	}
	res, err := http.Post(PUSHOVER_API_URL, "application/json", bytes.NewBuffer(jsonBody))
	if err != nil {
		return err
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("pushover: %s", res.Status)
	}
	return nil
}
